File: coresuite-dotnet-framework-console-vb/merger/MergerInvoice.vb
Imports ceTe.DynamicPDF
Imports ceTe.DynamicPDF.Merger
Imports ceTe.DynamicPDF.PageElements
Imports ceTe.DynamicPDF.PageElements.BarCoding

Public Class MergerInvoice
    Public Shared order As List(Of MergerInvoiceData.Order) = MergerInvoiceData.Orders

    Friend Shared Function Run(invoiceNumbers() As String) As Byte()
        ' Create a document And set it's properties
        Dim Document As Document = New Document With {
        .Creator = "Invoice",
        .Author = "ceTe Software",
        .Title = "Invoice"
    }

        ' Add Invoices to the document
        Dim item As String
        Dim data As MergerInvoiceData.Order
        For Each item In invoiceNumbers
            For Each data In order
                If item = data.OrderID.ToString() Then
                    Dim Invoice As MergerInvoiceCreator = New MergerInvoiceCreator()
                    Invoice.Draw(data, Document)
                End If
            Next
        Next

        ' Outputs the Invoices to the current web page
        Try
            Return Document.Draw()
        Catch Empty As EmptyDocumentException
            ' Show an empty document error
            Return ShowErrorDocument("Document is empty. No invoices were selected.", "")
        Catch excGeneral As Exception
            'Show a general error
            Return ShowErrorDocument(excGeneral.Message, excGeneral.StackTrace)
        End Try
    End Function

    Private Shared Function ShowErrorDocument(message As String, stackTrace As String) As Byte()
        ' Shows the error in a PDF document
        Dim document As Document = New Document()
        Dim page As Page = New Page(PageSize.Letter, PageOrientation.Portrait)

        Dim messageArea As TextArea = New TextArea(message, 0, 0, 512, 36, Font.Helvetica, 18, RgbColor.Red)
        messageArea.Height = messageArea.GetRequiredHeight()
        page.Elements.Add(messageArea)

        If stackTrace.Trim().Length > 0 Then
            Dim stackTraceTop As Single = messageArea.Y + messageArea.Height + 20

            Dim stackTraceArea As TextArea = New TextArea(stackTrace, 0, stackTraceTop, 512, 692 - stackTraceTop, Font.Helvetica, 10)

            page.Elements.Add(New Label("StackTrace:", 0, stackTraceTop - 12, 512, 12, Font.HelveticaBold, 10))
            page.Elements.Add(stackTraceArea)
        End If

        document.Pages.Add(page)
        Return document.Draw()
    End Function
End Class

Public Class MergerInvoiceCreator

    Private Shared templatePath As String
    Private subTotal As Decimal = 0
    Private freight As Decimal = 0
    Private Shared templatePathSet As Boolean = False
    Public Sub New()
        SetTemplatePath()
    End Sub

    Private Sub SetTemplatePath()
        ' Adds the image to page template if it Is Not already added
        If Not templatePathSet Then
            templatePath = Util.GetResourcePath("PDFs/Invoice.pdf")
            templatePathSet = True
        End If
    End Sub

    Public Sub Draw(ByVal Data As MergerInvoiceData.Order, ByVal MyDocument As Document)
        ' Each Invoice should begin a New section
        MyDocument.Sections.Begin()

        ' Adds the invoice if there Is data
        If Not Data Is Nothing Then
            ' Sets the Freight amount
            freight = CType(Data.Freight, Decimal)

            'Draws the invoice data, returns a page object if it Is the last page
            Dim i As Integer = 0
            Dim LastPage As Page = DrawInvoiceData(Data, MyDocument, i)

            ' Draws aditional pages if necessary
            While LastPage Is Nothing
                LastPage = DrawInvoiceData(Data, MyDocument, i)
            End While
            'Draws the totals to the bottom of the last page of the Invoice
            DrawTotals(Data, MyDocument.Pages.Item(MyDocument.Pages.Count - 1))
        End If
    End Sub

    Private Function DrawInvoiceData(ByVal Data As MergerInvoiceData.Order, ByVal MyDocument As Document, ByRef start As Integer) As Page
        'Tracks if the invoice is finished
        Dim InvoiceFinished As Boolean = True
        'Tracks the y position on the MyPage
        Dim YOffset As Single = 288

        'Create a MyPage for the Invoice
        Dim MyPage As Page = New ImportedPage(templatePath, 1)
        'Add Details to the Invoice
        DrawInvoiceDetails(Data, MyPage)
        'Add bill to address
        MyPage.Elements.Add(New TextArea(Data.BillTo, 28, 138, 194, 70, Font.Helvetica, 12))
        'Add ship to address
        MyPage.Elements.Add(New TextArea(Data.ShipTo, 318, 138, 194, 70, Font.Helvetica, 12))

        'Add line items to the Invoice
        For i As Integer = start To Data.OrderDetails.Count - 1
            DrawLineItem(Data.OrderDetails.Item(i), MyPage, YOffset)
            'Break if at the bottom of the MyPage
            If YOffset >= 666 Then
                InvoiceFinished = False
                start = i + 1
                Exit For
            End If
        Next i

        'Add the MyPage to the MyDocument
        MyDocument.Pages.Add(MyPage)

        'If Invoice is finished return the MyPage else return null so another MyPage will be added
        If InvoiceFinished Then
            Return MyPage
        Else
            MyPage.Elements.Add(New Label("Continued...", 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right))
            Return Nothing
        End If
    End Function

    Private Sub DrawInvoiceDetails(ByVal Data As MergerInvoiceData.Order, ByVal MyPage As Page)
        'Adds Invoice details to the MyPage
        MyPage.Elements.Add(New Label(Data.OrderID.ToString(), 437, 25, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label((CType(Data.OrderDate, DateTime)).ToString("d MMM yyyy"), 437, 39, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label(Data.CustomerID.ToString(), 437, 53, 100, 12, Font.Helvetica, 12))
        If Not Data.ShippedDate = Nothing Then
            MyPage.Elements.Add(New Label((CType(Data.ShippedDate, DateTime)).ToString("d MMM yyyy"), 437, 67, 100, 12, Font.Helvetica, 12))
        End If
        MyPage.Elements.Add(New Label(Data.ShipperName.ToString(), 437, 81, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Interleaved25(Data.OrderID.ToString(), 380, 4, 18, False))
        MyPage.Elements.Add(New PageNumberingLabel("Page %%SP%% of %%ST%% ", 450, 253, 90, 20, Font.HelveticaBold, 12, TextAlign.Center))
    End Sub

    Private Sub DrawLineItem(ByVal Data As MergerInvoiceData.OrderDetail, ByVal MyPage As Page, ByRef YOffset As Single)
        'Adds a line item to the invoice
        Dim Quantity As Int16 = CType(Data.Quantity, Int16)
        Dim UnitPrice As Decimal = CType(Data.UnitPrice, Decimal)
        Dim LineTotal As Decimal = UnitPrice * Quantity
        subTotal += LineTotal

        MyPage.Elements.Add(New Label(Quantity.ToString(), 4, 3 + YOffset, 52, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(Data.ProductName.ToString(), 64, 3 + YOffset, 292, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label(UnitPrice.ToString("#,##0.00"), 364, 3 + YOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(LineTotal.ToString("#,##0.00"), 454, 3 + YOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right))

        YOffset += 18
    End Sub

    Private Sub DrawTotals(ByVal Data As MergerInvoiceData.Order, ByVal MyPage As Page)
        'Add totals to the bottom of the Invoice
        Dim GrandTotal As Decimal = subTotal + freight

        MyPage.Elements.Add(New Label(subTotal.ToString("#,##0.00"), 454, 668, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(freight.ToString("#,##0.00"), 454, 686, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(GrandTotal.ToString("#,##0.00"), 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right))
    End Sub
End Class

File: coresuite-dotnet-core-console-cs/merger/MergerInvoice.cs
using ceTe.DynamicPDF;
using ceTe.DynamicPDF.Merger;
using ceTe.DynamicPDF.PageElements;
using ceTe.DynamicPDF.PageElements.BarCoding;
using System;
using System.Collections.Generic;

namespace coresuite_dotnet_core_console_cs.merger
{
    class MergerInvoice
    {
        public static List<MergerInvoiceData.Order> order = MergerInvoiceData.Orders;

        internal static byte[] Run(string[] invoiceNumbers)
        {
            // Create a document and set it's properties
            Document document = new Document
            {
                Creator = "Invoice",
                Author = "ceTe Software",
                Title = "Invoice"
            };

            // Add Invoices to the document
            foreach (string item in invoiceNumbers)
            {
                foreach (MergerInvoiceData.Order data in order)
                {
                    if (item == data.OrderID.ToString())
                    {
                        MergerInvoiceCreator invoice = new MergerInvoiceCreator();
                        invoice.Draw(data, document);
                    }
                }
            }
          
            // Outputs the Invoices to the current web page
            try
            {
                return document.Draw();
            }
            catch (EmptyDocumentException)
            {
                // Show an empty document error
                return ShowErrorDocument("Document is empty. No invoices were selected.", "");
            }
            catch (Exception excGeneral)
            {
                // Show a general error
                return ShowErrorDocument(excGeneral.Message, excGeneral.StackTrace);
            }
        }

        private static byte[] ShowErrorDocument(string message, string stackTrace)
        {
            // Shows the error in a PDF document
            Document document = new Document();
            Page page = new Page(PageSize.Letter, PageOrientation.Portrait);

            TextArea messageArea = new TextArea(message, 0, 0, 512, 36, Font.Helvetica, 18, RgbColor.Red);
            messageArea.Height = messageArea.GetRequiredHeight();
            page.Elements.Add(messageArea);

            if (stackTrace.Trim().Length > 0)
            {
                float stackTraceTop = messageArea.Y + messageArea.Height + 20;

                TextArea stackTraceArea = new TextArea(stackTrace, 0, stackTraceTop, 512, 692 - stackTraceTop, Font.Helvetica, 10);

                page.Elements.Add(new Label("StackTrace:", 0, stackTraceTop - 12, 512, 12, Font.HelveticaBold, 10));
                page.Elements.Add(stackTraceArea);
            }
            document.Pages.Add(page);
            return document.Draw();
        }
    }

    public class MergerInvoiceCreator
    {
        #region Member Variables

        private static string templatePath;
        private decimal subTotal = 0;
        private decimal freight = 0;
        private static bool templatePathSet = false;

        #endregion

        #region Constructors

        public MergerInvoiceCreator()
        {
            // Adds the image to page template if it is not already added
            if (!templatePathSet)
            {
                templatePath = Util.GetResourcePath("PDFs/Invoice.pdf");
                templatePathSet = true;
            }
        }

        #endregion

        #region Private Methods

        public void Draw(MergerInvoiceData.Order data, Document document)
        {
            // Each Invoice should begin a new section
            document.Sections.Begin();

            // Adds the invoice if there is data
            if (data != null)
            { 
                // Sets the Freight amount
                freight = (decimal)data.Freight;

                // Draws the invoice data, returns a page object if it is the last page
                int i = 0;
                Page lastPage = DrawInvoiceData(data, document,ref i);

                // Draws aditional pages if necessary
                while (lastPage == null)
                {
                    lastPage = DrawInvoiceData(data, document,ref i);
                }

                // Draws the totals to the bottom of the last page of the Invoice
                DrawTotals(data, document.Pages[document.Pages.Count - 1]);
            }
        }

        private Page DrawInvoiceData(MergerInvoiceData.Order data, Document document,ref int start)
        {
            // Tracks if the invoice is finished
            bool invoiceFinished = true;

            // Tracks the y position on the page
            float yOffset = 288;

            // Create a page using "Invoice.pdf" as a template
            Page page = new ImportedPage(templatePath, 1);

            // Uncomment the line below to show a layout grid.
            //page.Elements.Add( new LayoutGrid() );

            // Add Details to the Invoice
            DrawInvoiceDetails(data, page);

            // Add bill to address         
            page.Elements.Add(new TextArea(data.BillTo, 28, 139, 194, 70, Font.Helvetica, 12));

            // Add ship to address           
            page.Elements.Add(new TextArea(data.ShipTo, 318, 139, 194, 70, Font.Helvetica, 12));

            // Add line items to the Invoice   
            for (int i = start; i < data.OrderDetails.Count; i++)
            {
                DrawLineItem(data.OrderDetails[i], page, ref yOffset);
                // Break if at the bottom of the page
                if (yOffset >= 666)
                {
                    invoiceFinished = false;
                    start = i + 1;
                    break;
                }
            }

            // Add the page to the document
            document.Pages.Add(page);

            // If Invoice is finished return the page else return null so another page will be added
            if (invoiceFinished)
            {
                return page;
            }
            else
            {
                page.Elements.Add(new Label("Continued...", 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                return null;
            }
        }

        private void DrawInvoiceDetails(MergerInvoiceData.Order data, Page page)
        {
            // Adds Invoice details to the page
            page.Elements.Add(new Label(data.OrderID.ToString(), 437, 25, 100, 12, Font.Helvetica, 12));
            page.Elements.Add(new Label(((DateTime)data.OrderDate).ToString("d MMM yyyy"), 437, 39, 100, 12, Font.Helvetica, 12));
            page.Elements.Add(new Label(data.CustomerID.ToString(), 437, 53, 100, 12, Font.Helvetica, 12));
            if (!(data.ShippedDate==null))
            {
                page.Elements.Add(new Label(((DateTime)data.ShippedDate).ToString("d MMM yyyy"), 437, 67, 100, 12, Font.Helvetica, 12));
            }
            page.Elements.Add(new Label(data.ShipperName.ToString(), 437, 81, 100, 12, Font.Helvetica, 12));
            page.Elements.Add(new Interleaved25(data.OrderID.ToString(), 380, 4, 18, false));
            page.Elements.Add(new PageNumberingLabel("Page %%SP%% of %%ST%% ", 450, 253, 90, 20, Font.HelveticaBold, 12, TextAlign.Center));
        }

        private void DrawLineItem(MergerInvoiceData.OrderDetail detail, Page page, ref float yOffset)
        {
            // Adds a line item to the invoice
            Int16 quantity = (Int16)detail.Quantity;
            decimal unitPrice = (decimal)detail.UnitPrice;
            decimal lineTotal = unitPrice * quantity;
            subTotal += lineTotal;

            page.Elements.Add(new Label(quantity.ToString(), 4, 3 + yOffset, 52, 12, Font.Helvetica, 12, TextAlign.Right));
            page.Elements.Add(new Label(detail.ProductName.ToString(), 64, 3 + yOffset, 292, 12, Font.Helvetica, 12));
            page.Elements.Add(new Label(unitPrice.ToString("#,##0.00"), 364, 3 + yOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right));
            page.Elements.Add(new Label(lineTotal.ToString("#,##0.00"), 454, 3 + yOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right));
            yOffset += 18;
        }

        private void DrawTotals(MergerInvoiceData.Order data, Page page)
        {
            // Add totals to the bottom of the Invoice
            decimal grandTotal = subTotal + freight;
            page.Elements.Add(new Label(subTotal.ToString("#,##0.00"), 454, 668, 82, 12, Font.Helvetica, 12, TextAlign.Right));
            page.Elements.Add(new Label(freight.ToString("#,##0.00"), 454, 686, 82, 12, Font.Helvetica, 12, TextAlign.Right));
            page.Elements.Add(new Label(grandTotal.ToString("#,##0.00"), 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right));
        }
       #endregion
    }
}
